File: Assets/Scripts/Core/Testing/HeadlessGameRunner.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Bunker.Content;
using UnityEngine;

namespace Bunker.Core.Testing
{
    // Runs a full game session from start to finish without any UI or real time,
    // driving GameSession purely through its public API by simulating random
    // (but valid) player decisions. Used both as a manual debugging tool and
    // as the foundation for automated unit tests (Unity Test Framework).
    public class HeadlessGameRunner
    {
        private readonly System.Random _decisionRandom;
        public List<string> Log { get; } = new();

        public HeadlessGameRunner(int? decisionSeed = null)
        {
            _decisionRandom = decisionSeed.HasValue
                ? new System.Random(decisionSeed.Value)
                : new System.Random();
        }

        // Builds a fresh session with fakePlayerCount players and runs it to completion.
        // Returns the final GameOverResult for assertions in tests.
        public GameOverResult RunFullSimulation(
            int fakePlayerCount,
            List<TraitPoolSO> traitPools,
            SpecialCardPoolSO specialCardPool,
            int survivorsTarget,
            int? dealSeed = null)
        {
            var players = CreateFakePlayers(fakePlayerCount);
            var config = new GameSessionConfig(survivorsTarget, dealSeed);
            var generator = new CharacterCardGenerator(traitPools, specialCardPool, dealSeed);
            var session = new GameSession(players, config, generator);

            GameOverResult finalResult = null;

            // Wire up logging + capture the terminal result via events,
            // mirroring how a real UI/network layer would react to GameSession.
            session.OnPhaseChanged += phase => LogLine($"Phase -> {phase}");
            session.OnRoundStarted += round => LogLine($"--- Round {round} started ---");
            session.OnTraitRevealed += (player, trait) =>
                LogLine($"{player.DisplayName} revealed {trait.Category}: {trait.Id}");
            session.OnSpecialCardUsed += (player, card) =>
                LogLine($"{player.DisplayName} used special card: {card.Id} ({card.EffectType})");
            session.OnVotingResolved += result =>
                LogLine($"Vote resolved: {result.ResultType} " +
                         $"({string.Join(", ", result.VoteCounts.Select(kv => $"{kv.Key}={kv.Value}"))})");
            session.OnPlayerEliminated += player =>
                LogLine($"ELIMINATED: {player.DisplayName}");
            session.OnGameOverResolved += result =>
            {
                finalResult = result;
                LogLine($"GAME OVER: {result.EndReason}, survivors: " +
                         string.Join(", ", result.Survivors.Select(p => p.DisplayName)));
            };

            var validation = session.ValidateCanStart();
            if (!validation.CanStart)
            {
                LogLine($"START FAILED: {validation.FailReason}");
                return null;
            }

            session.StartGame();
            DriveGameToCompletion(session);

            return finalResult;
        }

        // --- Player simulation ---------------------------------------------

        private List<PlayerData> CreateFakePlayers(int count)
        {
            var players = new List<PlayerData>();
            for (int i = 0; i < count; i++)
            {
                players.Add(new PlayerData($"fake_player_{i}", $"Player {i}"));
            }
            return players;
        }

        // Drives the session through Reveal -> Discussion -> Voting (-> Tiebreaker)
        // -> next round, repeatedly, until GameOver. Safety-capped to avoid an
        // infinite loop if a rules bug causes the game to never end.
        private void DriveGameToCompletion(GameSession session)
        {
            const int maxIterations = 1000;
            int iterations = 0;

            while (session.Phase != GamePhase.GameOver && iterations < maxIterations)
            {
                iterations++;

                switch (session.Phase)
                {
                    case GamePhase.Reveal:
                        SimulateRevealPhase(session);
                        break;

                    case GamePhase.Discussion:
                        // No real timer here — immediately proceed to voting,
                        // since headless runs don't need to wait real seconds.
                        session.StartVotingPhase();
                        break;

                    case GamePhase.Voting:
                    case GamePhase.VotingTiebreaker:
                        SimulateVotingPhase(session);
                        break;

                    case GamePhase.RoundResult:
                        // GameSession auto-advances from RoundResult internally
                        // (FinishRound -> StartNextRound or EndGame), so this
                        // phase should be transient. If we ever observe it here,
                        // something didn't advance — treat as a bug signal.
                        LogLine("WARNING: observed RoundResult phase directly, possible logic gap.");
                        iterations = maxIterations; // bail out
                        break;

                    default:
                        // Lobby / Dealing should never be seen mid-loop.
                        LogLine($"WARNING: unexpected phase in drive loop: {session.Phase}");
                        iterations = maxIterations;
                        break;
                }
            }

            if (iterations >= maxIterations)
            {
                LogLine("ABORTED: exceeded max iterations, possible infinite loop in game rules.");
            }
        }

        private void SimulateRevealPhase(GameSession session)
        {
            // Keep asking the current turn player to reveal traits/use their
            // special card until a full pass completes and phase changes away
            // from Reveal (StartDiscussionPhase is called by the caller,
            // simulated here right after the pass completes).
            bool passCompleted = false;
            session.OnRevealPassCompleted += LocalHandler;
            void LocalHandler() => passCompleted = true;

            int safety = 0;
            while (session.Phase == GamePhase.Reveal && safety < 500)
            {
                safety++;
                var playerId = session.CurrentTurnPlayerId;
                if (playerId == null) break;

                var player = session.GetPlayer(playerId);

                // Small chance to use the special card before revealing, if available.
                if (!player.HasUsedSpecialCard && player.Special != null && RollChance(0.15))
                {
                    TrySimulateSpecialCardUse(session, player);
                }

                var hiddenCategory = player.Traits
                    .Select(t => t.Category)
                    .FirstOrDefault(c => !player.IsCategoryRevealed(c));

                session.RevealNextTrait(playerId, hiddenCategory);

                if (passCompleted)
                {
                    session.OnRevealPassCompleted -= LocalHandler;
                    session.StartDiscussionPhase(durationSeconds: 60);
                    break;
                }
            }

            session.OnRevealPassCompleted -= LocalHandler;
        }

        private void TrySimulateSpecialCardUse(GameSession session, PlayerData caster)
        {
            var possibleTargets = session.ActivePlayers()
                .Where(p => p.PlayerId != caster.PlayerId)
                .ToList();

            if (possibleTargets.Count == 0) return;

            var target = possibleTargets[_decisionRandom.Next(possibleTargets.Count)];

            if (caster.Special.EffectType == SpecialCardEffectType.SwapTrait)
            {
                var swappableCategory = caster.Traits
                    .Select(t => t.Category)
                    .FirstOrDefault(c => !caster.IsCategoryRevealed(c) && !target.IsCategoryRevealed(c));

                if (swappableCategory != default)
                {
                    session.UseSwapTraitSpecialCard(caster.PlayerId, target.PlayerId, swappableCategory);
                }
            }
            else
            {
                session.UseSpecialCard(caster.PlayerId, target.PlayerId);
            }
        }

        private void SimulateVotingPhase(GameSession session)
        {
            foreach (var voter in session.ActivePlayers().ToList())
            {
                var possibleTargets = session.ActivePlayers()
                    .Where(p => p.PlayerId != voter.PlayerId && !p.HasVoteImmunityThisRound)
                    .ToList();

                if (possibleTargets.Count == 0) continue;

                var target = possibleTargets[_decisionRandom.Next(possibleTargets.Count)];
                session.CastVote(voter.PlayerId, target.PlayerId);
            }

            session.ResolveVotes();
        }

        private bool RollChance(double probability)
        {
            return _decisionRandom.NextDouble() < probability;
        }

        private void LogLine(string line)
        {
            Log.Add(line);
        }
    }
}

File: bunker-server/src/game/testing/HeadlessGameRunner.ts
import { PlayerData } from "../PlayerData";
import { GameSession } from "../GameSession";
import { GameSessionConfig } from "../GameSessionConfig";
import { CharacterCardGenerator } from "../content/CharacterCardGenerator";
import { TraitPool } from "../content/TraitPool";
import { SpecialCardPool } from "../content/SpecialCardPool";
import { GamePhase } from "../GamePhase";
import { GameOverResult } from "../GameOverResult";
import { SeededRandom } from "../content/SeededRandom";
import { SpecialCardEffectType } from "../SpecialCard";

// Direct port of the C# HeadlessGameRunner (section 1.8). Drives a
// GameSession through a full game using randomized but valid decisions,
// without any real time or network - used both as a manual debugging tool
// and as the basis for automated tests.
export class HeadlessGameRunner {
  private readonly decisionRandom: SeededRandom;
  public readonly log: string[] = [];

  constructor(decisionSeed?: number) {
    this.decisionRandom = new SeededRandom(decisionSeed);
  }

  runFullSimulation(
    fakePlayerCount: number,
    traitPools: TraitPool[],
    specialCardPool: SpecialCardPool,
    survivorsTarget: number,
    dealSeed?: number
  ): GameOverResult | null {
    const players = this.createFakePlayers(fakePlayerCount);
    const config: GameSessionConfig = { survivorsTarget, randomSeed: dealSeed, allowVoteTies: false };
    const generator = new CharacterCardGenerator(traitPools, specialCardPool, dealSeed);
    const session = new GameSession(players, config, generator);

    let finalResult: GameOverResult | null = null;

    session.on("phaseChanged", (phase) => this.logLine(`Phase -> ${GamePhase[phase]}`));
    session.on("roundStarted", (round) => this.logLine(`--- Round ${round} started ---`));
    session.on("traitRevealed", (player, trait) =>
      this.logLine(`${player.displayName} revealed ${trait.category}: ${trait.id}`)
    );
    session.on("specialCardUsed", (player, card) =>
      this.logLine(`${player.displayName} used special card: ${card.id} (${card.effectType})`)
    );
    session.on("votingResolved", (result) => {
      const counts = [...result.voteCounts.entries()].map(([id, c]) => `${id}=${c}`).join(", ");
      this.logLine(`Vote resolved: ${result.resultType} (${counts})`);
    });
    session.on("playerEliminated", (player) => this.logLine(`ELIMINATED: ${player.displayName}`));
    session.on("gameOverResolved", (result) => {
      finalResult = result;
      const names = result.survivors.map((p) => p.displayName).join(", ");
      this.logLine(`GAME OVER: ${result.endReason}, survivors: ${names}`);
    });

    const validation = session.validateCanStart();
    if (!validation.canStart) {
      this.logLine(`START FAILED: ${validation.failReason}`);
      return null;
    }

    session.startGame();
    this.driveGameToCompletion(session);

    return finalResult;
  }

  // --- Player simulation ---------------------------------------------

  private createFakePlayers(count: number): PlayerData[] {
    return Array.from({ length: count }, (_, i) => new PlayerData(`fake_player_${i}`, `Player ${i}`));
  }

  // Drives the session through Reveal -> Discussion -> Voting (-> Tiebreaker)
  // -> next round, repeatedly, until GameOver. Safety-capped to avoid an
  // infinite loop if a rules bug causes the game to never end.
  private driveGameToCompletion(session: GameSession): void {
    const maxIterations = 1000;
    let iterations = 0;

    while (session.phase !== GamePhase.GameOver && iterations < maxIterations) {
      iterations++;

      switch (session.phase) {
        case GamePhase.Reveal:
          this.simulateRevealPhase(session);
          break;

        case GamePhase.Discussion:
          // No real timer - immediately proceed to voting for headless runs.
          session.startVotingPhase();
          break;

        case GamePhase.Voting:
        case GamePhase.VotingTiebreaker:
          this.simulateVotingPhase(session);
          break;

        case GamePhase.RoundResult:
          // GameSession auto-advances from RoundResult internally
          // (finishRound -> startNextRound or endGame). Observing it here
          // directly means something didn't advance - treat as a bug signal.
          this.logLine("WARNING: observed RoundResult phase directly, possible logic gap.");
          iterations = maxIterations;
          break;

        default:
          this.logLine(`WARNING: unexpected phase in drive loop: ${GamePhase[session.phase]}`);
          iterations = maxIterations;
          break;
      }
    }

    if (iterations >= maxIterations) {
      this.logLine("ABORTED: exceeded max iterations, possible infinite loop in game rules.");
    }
  }

  private simulateRevealPhase(session: GameSession): void {
    let passCompleted = false;
    const handler = () => {
      passCompleted = true;
    };
    session.on("revealPassCompleted", handler);

    let safety = 0;
    while (session.phase === GamePhase.Reveal && safety < 500) {
      safety++;
      const playerId = session.currentTurnPlayerId;
      if (!playerId) break;

      const player = session.getPlayer(playerId)!;

      // Small chance to use the special card before revealing, if available.
      if (!player.hasUsedSpecialCard && player.special && this.rollChance(0.15)) {
        this.trySimulateSpecialCardUse(session, player);
      }

      const hiddenCategory = player.getTraits().find((t) => !player.isCategoryRevealed(t.category))?.category;
      if (hiddenCategory !== undefined) {
        session.revealNextTrait(playerId, hiddenCategory);
      }

      if (passCompleted) {
        session.off("revealPassCompleted", handler);
        session.startDiscussionPhase(60);
        return;
      }
    }

    session.off("revealPassCompleted", handler);
  }

  private trySimulateSpecialCardUse(session: GameSession, caster: PlayerData): void {
    const possibleTargets = session.activePlayers().filter((p) => p.playerId !== caster.playerId);
    if (possibleTargets.length === 0) return;

    const target = possibleTargets[this.decisionRandom.nextInt(possibleTargets.length)];

    if (caster.special!.effectType === SpecialCardEffectType.SwapTrait) {
      const swappableCategory = caster
        .getTraits()
        .find((t) => !caster.isCategoryRevealed(t.category) && !target.isCategoryRevealed(t.category))?.category;

      if (swappableCategory !== undefined) {
        session.useSwapTraitSpecialCard(caster.playerId, target.playerId, swappableCategory);
      }
    } else {
      session.useSpecialCard(caster.playerId, target.playerId);
    }
  }

  private simulateVotingPhase(session: GameSession): void {
    for (const voter of session.activePlayers()) {
      const possibleTargets = session
        .activePlayers()
        .filter((p) => p.playerId !== voter.playerId && !p.hasVoteImmunityThisRound);

      if (possibleTargets.length === 0) continue;

      const target = possibleTargets[this.decisionRandom.nextInt(possibleTargets.length)];
      session.castVote(voter.playerId, target.playerId);
    }

    session.resolveVotes();
  }

  private rollChance(probability: number): boolean {
    return this.decisionRandom.nextFloat() < probability;
  }

  private logLine(line: string): void {
    this.log.push(line);
  }
}
